### ISSUE TITLE
Add Support for GDSII Files

### ISSUE THREAD

**Issue by Monkeymerlot** [levylabpitt/AFM-Lithography#59]
Currently we use .svg (scalable vector graphics) files. However, with the addition of E-Beam Lithography to our arsenal, which uses the industry standard GDSII files, we would like to add support for GDSII files. Previous discussion is in Issue #16.

**Comment by Monkeymerlot** [comment]
@ciozi137 Until we figure out how we are going to structure the data and how it will fit in with the Lithography program overall, I am going to build a GDSII viewer. The viewer part of it will be really easy to implement and it will let us check our work.

**Comment by Monkeymerlot** [comment]
https://www.artwork.com/gdsii/gdsii/page5.htm
https://boolean.klaasholwerda.nl/interface/bnf/gdsformat.html

**Comment by Monkeymerlot** [comment]
@ciozi137 I am having trouble with some number conversions (probably because LabVIEW is 32 bit and this number is 64 bits). There are two numbers as explained [here.](https://boolean.klaasholwerda.nl/interface/bnf/gdsformat.html#rec_units)
Further up in the [page](https://boolean.klaasholwerda.nl/interface/bnf/gdsformat.html#datatypes) there is also an explanation about how to read them. 

I wrote a test and everything for you if you want to take a crack at it...

**Comment by ciozi137** [comment]
Which recordtype code are you having trouble with? I can look into this more tomorrow. Also point me to your test files

**Comment by Monkeymerlot** [comment]
I forgot to attach the test files! And the Unit Record type right now but this is used for other important record types as well.
[GDSII.zip](https://github.com/levylabpitt/AFM-Lithography/files/5915949/GDSII.zip)

**Comment by ciozi137** [comment]
Agree it's good to work this out on one datatype first. I don't think 32/64 bit labview should be an issue

**Comment by Monkeymerlot** [comment]
From https://github.com/heitzmann/gdspy:
(https://github.com/heitzmann/gdspy/blob/master/gdspy/gdsiiformat.py)
```
def _eight_byte_real_to_float(value):
    """
    Convert a number from GDSII 8 byte real format to float.
    Parameters
    ----------
    value : string
        The GDSII binary string representation of the number.
    Returns
    -------
    out : float
        The number represented by `value`.
    """
    short1, short2, long3 = struct.unpack(">HHL", value)
    exponent = (short1 & 0x7F00) // 256 - 64
    mantissa = (
        ((short1 & 0x00FF) * 65536 + short2) * 4294967296 + long3
    ) / 72057594037927936.0
    if short1 & 0x8000:
        return -mantissa * 16.0 ** exponent
    return mantissa * 16.0 ** exponent

```

I did some more searching and it seems to be that this is not even a normal standard format.

**Comment by ciozi137** [comment]
@Monkeymerlot I think I got it! I'll send the code back to you for you to look at. Here are a few comments:
1. The string data didn't make any sense to me with respect to the GDSII description, so I read as byte array (I think this also makes parsing faster later)

![image](https://user-images.githubusercontent.com/26659428/106798689-54aab580-662c-11eb-84c0-163cf412df83.png)

2. I separate Record Type and Data Type bytes. For some reason the reference you linked above adds the two bytes together, but this doesn't make sense to me because there are only 7 datatypes, not 44+

![image](https://user-images.githubusercontent.com/26659428/106798717-5e341d80-662c-11eb-9f9f-678e81961dcd.png)

3. I think you were on the right track with the conversion for 8 byte float, but it is a number not an array. The floating point conversion you linked from gdspy library is normal. I don't know why they implemented that routine from scratch in Python. 

![image](https://user-images.githubusercontent.com/26659428/106798757-6ee49380-662c-11eb-8d10-0817d56c0f71.png)

I used [this ](https://jupiter.math.nctu.edu.tw/~weng/courses/IC_2007/PROJECT_NCTU_MATH/CELL_LAYOUT/The%20GDSII%20Stream%20Format.htm) as my reference for Reference and Data Types. There were more than you had and some of the numbers were different.

**Comment by ciozi137** [comment]
[GDSII-patrick.zip](https://github.com/levylabpitt/AFM-Lithography/files/5921025/GDSII-patrick.zip)

**Comment by Monkeymerlot** [comment]
Sweet! And there are 44 record types, do you can combine the record type with the data type into one code. I was thinking that it would help in the future for structuring the data but your way looks much cleaner.

**Comment by ciozi137** [comment]
The other thing I forgot to mention is with reading binary file you can control endianness, but I don't think that ended up being necessary.

**Comment by Monkeymerlot** [comment]
Ok so I finally got a chance to look at the VI you sent (I was cooking stuff) and the only issue is that the value for `UNITS` is still incorrect. It should be two values, the first of which is 0.001 and the second is 1e-9. The format for the GDSII file 8 byte real is (sign)* mantissa * 16^exponent, while the exponent uses excess 64 notation, while the format for normal double precision (IEEE-754) is (sign) * Mantissa * 2^exponent, where the exponent is in offset binary representation.

**Comment by ciozi137** [comment]
OK I will look again.

**Comment by ciozi137** [comment]
@Monkeymerlot ok so there **is** something called "Excess-64" representation. I tried to get it working (see attached; it works for some 4-byte-test cases), but...tag you're it.
[Patrick Excess-64.zip](https://github.com/levylabpitt/AFM-Lithography/files/5922085/Patrick.Excess-64.zip)

**Comment by Monkeymerlot** [comment]
OK thanks for the help. Who would have guessed the very thing that has kept gdsii relevant (defining its own datatypes, well defined structure) would be what is holding us back 40 years later...

**Comment by ciozi137** [comment]
@Monkeymerlot see if you can understand what I coded. I tried to add enough comments to make it comprehensible. I think there is something obscure like a MSB/LSB issue. Also I think the Excess-64 is standard for CS people

**Comment by Monkeymerlot** [comment]
From https://jupiter.math.nctu.edu.tw/~weng/courses/IC_2007/PROJECT_NCTU_MATH/CELL_LAYOUT/The%20GDSII%20Stream%20Format.htm
```
The sixth data type is the "Eight Byte Real". The code is 0x05. This one
gets a little more use.

The first (most significant) bit of the first byte is the sign, one means
negative, 0 means positive.

The 7 least significant bits of the first byte are the exponent in "excess
64" notation. You must subtract 64 to get the true value. I'll show the
subtraction in the formula below.

The remaining 7 bytes are the mantissa, with a binary point to the left of
the most significant figure. The formula below uses the unsigned integer
value of these 7 bytes as the numerator of a fraction.

value = (mantissa/(2^56)) * (16^(exponent-64))
```

Well using this source and what you had + some internet binary converters I was able to get the correct answer:
![image](https://user-images.githubusercontent.com/37567672/106823276-a49a7400-664e-11eb-9cdf-5bd076d61f52.png)
So I think that we are almost there...

**Comment by ciozi137** [comment]
@Monkeymerlot I should have pasted my results. It looks like I had some copy paste errors with my test numbers. 0 is still not working. And I think I still have issues converting the binary number from the file to bits in LV, but maybe you have that issue worked out:
![image](https://user-images.githubusercontent.com/26659428/106824808-94d05f00-6651-11eb-97f0-f3c79c637d0d.png)

**Comment by Monkeymerlot** [comment]
OK I think I just had a breakthrough so I will post something shortly...

**Comment by Monkeymerlot** [comment]
I finally got it! Wow this turned out to be way more difficult because of MSB/LSB (you were right).
![image](https://user-images.githubusercontent.com/37567672/106828762-d31d4c80-6658-11eb-9003-8a22ea15f2ef.png)
![image](https://user-images.githubusercontent.com/37567672/106828776-db758780-6658-11eb-86df-efd102251274.png)

**Comment by Monkeymerlot** [comment]
Actually that isn't quite right. There needs to be a reverse array after the delete from array. As luck would have it, the number was 0111110...

**Comment by ciozi137** [comment]
Oh nice I didn't realize you could set the output format of Boolean array to number. I think that is good to know and will be helpful.

**Comment by Monkeymerlot** [comment]
Before I figured it out I was just using a loop and manually finding the number like this:
![image](https://user-images.githubusercontent.com/37567672/106830896-f1854700-665c-11eb-894b-c1a84e386a34.png)
but was getting the same wrong answer... so then I realized it wasn't the format but probably the way that LabVIEW was reading it because I could also convert to a string in this loop, concatenate the string array after the loop, and then copy paste into an online binary to decimal converter. I was getting the right result online, but not in LabVIEW.

Once I got LabVIEW in check I went back and tried to find a way to output a bigger number and noticed the properties menu...

**Comment by ciozi137** [comment]
@Monkeymerlot I think I'm getting closer: [GDSII-patrick-2.zip](https://github.com/levylabpitt/AFM-Lithography/files/5923392/GDSII-patrick-2.zip)

![image](https://user-images.githubusercontent.com/26659428/106848059-28b82000-667e-11eb-8667-a4458963420d.png)

![image](https://user-images.githubusercontent.com/26659428/106848143-58ffbe80-667e-11eb-952a-a2ec63ae061e.png)

**Comment by ciozi137** [comment]
@Monkeymerlot now re-reading your [comment ](https://github.com/levylabpitt/AFM-Lithography/issues/59#issuecomment-772945091) from 5 hours ago...I can't tell if you were able to get everything working or not lol

**Comment by Monkeymerlot** [comment]
@ciozi137 Yes I was, my comment didn't seem to get posted... Sorry if you have been working on this all night...

Comment that didn't get posted:

Got it to work! Added for loop to account for multiple numbers if the array is bigger than 8 bytes as well:
![image](https://user-images.githubusercontent.com/37567672/106850119-edb7eb80-6681-11eb-97f3-699c87c525e3.png)

[ConvertByteArrayToGDSII8ByteReal.zip](https://github.com/levylabpitt/AFM-Lithography/files/5923459/ConvertByteArrayToGDSII8ByteReal.zip)

**Comment by ciozi137** [comment]
That's ok. Be careful here because a negative exponent is allowed, but you'll get something else if you don't convert to I8 first. 
![image](https://user-images.githubusercontent.com/26659428/106850699-02e14a00-6683-11eb-9bf0-be570b3e2254.jpeg)

Or see my post earlier

**Comment by Monkeymerlot** [comment]
@ciozi137 Yeah my window by my desk started dripping water from the snow so I guess in the midst of fixing that it didn't get posted...
![image](https://user-images.githubusercontent.com/37567672/106851063-a16dab00-6683-11eb-8714-8f306a4f30c5.png)
I have it set to i32 for the output, but it will always be a number between 0 and 127, which is why you have to subtract 64 from it (I just realized that is where the excess part comes from). I guess I got lucky because it's only 7 bits, so the first bit is positive (0) no matter what... Between that and the palindrome number in testing I got incredibly lucky with this, which is insane because I have been pulling my hair out over this on and off for like 5 days and almost missed it completely when I did get it...

**Comment by ciozi137** [comment]
@Monkeymerlot I had been keeping everything U64 to try to be consistent but then the exponent bias bit me because subtract won't convert to I64. I guess that datatype there is properly U8 --> I8

**Comment by ciozi137** [comment]
Hi @Monkeymerlot do you remember where we left off of with this task? It looks like we had stated some GDSII parsing but then got sidetracked with "Excess 64" discussion. ?

**Comment by ciozi137** [comment]
@Monkeymerlot this issue is resurrected. See latest commit in develop branch, in particular:
![image](https://user-images.githubusercontent.com/26659428/236318301-5195b113-7fdd-44de-88cb-7b6547ca836d.png)

**Comment by ciozi137** [comment]
@Monkeymerlot as a bridge to supporting GDSII, it may also be possible to export SVG files from KLayout
- https://www.klayout.de/forum/discussion/873/working-with-dxfs
